endpoint_route_handler/registry.py

```diff
@@ -11,103 +66,330 @@ class EndpointRegistry:
-            self._mapping[key] = rule
-            if not init:
-                self._refresh_update_required()
-            return True
-
-    def drop_rule(self, key):
-        existing = self._mapping.pop(key, None)
-        if not existing:
-            return False
-        self._refresh_update_required()
-        return True
+        )
+        if not cr.fetchone():
+            sql = """
+                CREATE SEQUENCE endpoint_route_version INCREMENT BY 1 START WITH 1;
+                CREATE OR REPLACE FUNCTION increment_endpoint_route_version()
+                    RETURNS TRIGGER AS $$
+                BEGIN
+                  PERFORM nextval('endpoint_route_version');
+                  RETURN NEW;
+                END;
+                $$ language plpgsql;
+                CREATE TRIGGER  update_endpoint_route_version_trigger
+                    BEFORE INSERT ON %(table)s
+                   for each row execute procedure increment_endpoint_route_version();
+                CREATE TRIGGER  insert_endpoint_route_version_trigger
+                    BEFORE UPDATE ON %(table)s
+                   for each row execute procedure increment_endpoint_route_version();
+            """
+            cr.execute(sql, {"table": AsIs(cls._table)})
 
-    def routing_update_required(self, http_id):
-        return http_id in self._http_ids_to_update
+    def __init__(self, cr):
+        self.cr = cr
 
-    def _refresh_update_required(self):
-        for http_id in self._http_ids:
-            self._http_ids_to_update.add(http_id)
+    def get_rules(self, keys=None, where=None):
+        for row in self._get_rules(keys=keys, where=where):
+            yield EndpointRule.from_row(self.cr.dbname, row)
 
-    def reset_update_required(self, http_id):
-        self._http_ids_to_update.discard(http_id)
+    def _get_rules(self, keys=None, where=None, one=False):
+        query = "SELECT * FROM endpoint_route"
+        pargs = ()
+        if keys and not where:
+            query += " WHERE key IN %s"
+            pargs = (tuple(keys),)
+        elif where:
+            query += " " + where
+        self.cr.execute(query, pargs)
+        return self.cr.fetchone() if one else self.cr.fetchall()
 
-    @classmethod
-    def registry_for(cls, dbname):
-        if dbname not in _REGISTRY_BY_DB:
-            _REGISTRY_BY_DB[dbname] = cls()
-        return _REGISTRY_BY_DB[dbname]
+    def _get_rule(self, key):
+        row = self._get_rules(keys=(key,), one=True)
+        if row:
+            return EndpointRule.from_row(self.cr.dbname, row)
 
-    @classmethod
-    def wipe_registry_for(cls, dbname):
-        if dbname in _REGISTRY_BY_DB:
-            del _REGISTRY_BY_DB[dbname]
+    def _lock_rows(self, keys):
+        sql = "SELECT id FROM endpoint_route WHERE key IN %s FOR UPDATE"
+        self.cr.execute(sql, (tuple(keys),), log_exceptions=False)
+
+    def _update(self, rows_mapping):
+        self._lock_rows(tuple(rows_mapping.keys()))
```

Comment:
Feedback: locking rows here is triggering a `psycopg2.extensions.TransactionRollbackError: could not serialize access due to concurrent update` error. Traceback is the following:
```python
[...]
  File "/odoo/src/odoo/__init__.py", line 104, in registry
    return modules.registry.Registry(database_name)
  File "/odoo/src/odoo/modules/registry.py", line 65, in __new__
    return cls.new(db_name)
  File "/odoo/src/odoo/modules/registry.py", line 89, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/odoo/src/odoo/modules/loading.py", line 571, in load_modules
    model._register_hook()
  File "/odoo/external-src/web-api/endpoint_route_handler/models/endpoint_route_sync_mixin.py", line 91, in _register_hook
    self.search(domain)._register_controllers(init=True)
  File "/odoo/external-src/wms/shopfloor_base/models/shopfloor_app.py", line 165, in _register_controllers
    super()._register_controllers(init=init, options=options)
  File "/odoo/external-src/web-api/endpoint_route_handler/models/endpoint_route_sync_mixin.py", line 102, in _register_controllers
    self._endpoint_registry.update_rules(rules, init=init)
  File "/odoo/external-src/web-api/endpoint_route_handler/registry.py", line 236, in update_rules
    self._update(to_update)
  File "/odoo/external-src/web-api/endpoint_route_handler/registry.py", line 200, in _update
    self._lock_rows(tuple(rows_mapping.keys()))
  File "/odoo/external-src/web-api/endpoint_route_handler/registry.py", line 197, in _lock_rows
    self.cr.execute(sql, (tuple(keys),), log_exceptions=False)
  File "<decorator-gen-3>", line 2, in execute
    
  File "/odoo/src/odoo/sql_db.py", line 101, in check
    return f(self, *args, **kwargs)
  File "/odoo/src/odoo/sql_db.py", line 298, in execute
    res = self._obj.execute(query, params)
psycopg2.extensions.TransactionRollbackError: could not serialize access due to concurrent update
```
This is happening when we start multiple Odoo instances at the same time (in our case it was 6), so each of them is trying to lock the same rows, generating this concurrency issue. Odoo is trying again & again to start until it successfully obtain the locks to do its stuff, and this is taking more and more time to boot depending on the number of instances to start.

We should try with a `pg_try_advisory_lock` or maybe (but I don't know the whole logic of the current code so tell me if it's possible):

- use of `FOR UPDATE SKIP LOCKED`, but we must return the ID of the rows which have been locked by doing so (`_lock_rows` has to returns these IDs)
- update only these IDs in the `query_multi_update` call. Lines which have been locked by another transaction will do the same, and if a transaction has no row to update it'll do nothing.

Now if the registry loading of an Odoo instance does nothing, is it an issue?

Reply:
IMO the loading of the registry should not modify the database. Only the update, install of new addon and regular operations on Odoo models should. :thinking:

Reply:
The table must be readable as soon as the routing map is generated. This happens as the 1st http request comes in. Hence, is pretty critical.

In general, I tend to say we should lock only the ones that are updated and in theory this is what should happen already.

Question: which records is trying to update? If you get this, I guess all. Which seems wrong.
By looking again at the hook, I'm not sure anymore that we should still call `register/unregister_controllers`.

The purpose of this was to make sure that all routes are correctly registered when the instance boots BUT since all routes are now stored in the DB the normal sync mechanism should take care of this when the instance is up and running.

Meaning: if you modify a route to be active/inactive, you should explicitly run the sync of the registry to register/unregister routes. There shouldn't be any reason to ensure again that they are all registered when it starts up.

And in this PR and in related PRs there's already an upgrade step that should register all existing route-dependent records.

I hope is clear :)

Reply:
I'm back (out of sickness finally :mechanical_arm:).
I've managed to reproduce in a dev env too.
Pushed a fix. I'll fwd port to it v16 later.